### ISSUE TITLE
Handling CQL3 collections

### ISSUE THREAD

**Issue by eplowe** [fluentcassandra/fluentcassandra#96]
Gonna add this one here as well. I pretty much use CQL 3 currently for all my cassandra stuff -- I'll tackle this with the other stuff I've committed too as well. 

http://www.datastax.com/dev/blog/cql3_collections

Need to implement the following Types:
- [ ] SetType - http://grepcode.com/file/repo1.maven.org/maven2/org.apache.cassandra/cassandra-all/1.2.0-beta1/org/apache/cassandra/db/marshal/SetType.java?av=f
- [ ] MapType - http://grepcode.com/file/repo1.maven.org/maven2/org.apache.cassandra/cassandra-all/1.2.0-beta1/org/apache/cassandra/db/marshal/MapType.java?av=f
- [ ] ListType - http://grepcode.com/file/repo1.maven.org/maven2/org.apache.cassandra/cassandra-all/1.2.0-beta1/org/apache/cassandra/db/marshal/MapType.java?av=f

**Comment by nberardi** [comment]
CQL 3 is now the default connection type as of the release of Cassandra 1.2, I don't know the fall out of this yet for people using the Cassandra 1.1 but I guess we will soon see.

**Comment by eplowe** [comment]
Excellent. Do you reckon the new collection data types in CQL3 will be
handled by FluentCassandra with no problem? That's why I put it as a place
holder issue for you to close or keep open as you deem necessary. I just
haven't had a chance to investigate been swamped :)

**Comment by nberardi** [comment]
But more to your point, I am guessing that Set will have to be a new Type under FluentCassandra.Types.

**Comment by eplowe** [comment]
Yup. As well as the map and list

**Comment by nberardi** [comment]
So we have the type names, now we just need to dig through the byte array and fit out how they are delimited.  I am betting the composite type is probably the best template for this, but it will take some digging through the source code to figure out what is actually happening.  

If you are interested in getting your feet wet before diving in there is a type called `EmptyType` which seems rather trivial to implement.  I referenced it in issue #97

**Comment by eplowe** [comment]
Excellent. I might actually wrap that up soon as I have some use cases that
I believe would be simplified in some existing services I maintain at my
day job.

**Comment by nberardi** [comment]
Remember if you decide to tackle this, these lists are generic types, and those generic types will be encoded as Big Endian, which is not as straight forward as you may think for all the types.  So make sure to reuse as much of the existing type system as already defined, and I think the `CompositeType` is going to be the best resource for these lists for a beginner to the complexity of the FluentCassandra type system.

**Comment by eplowe** [comment]
Understood! Thanks for the heads up. One of the things that drove me to FC
was the ease of dealing with the data types back and forth. You put a lot
of thought behind them and with that the codebase for the type system is
pretty very easy to follow.

**Comment by eplowe** [comment]
I have ListType and SetType working -- MapType should be done shortly.

**Comment by nberardi** [comment]
Is there anywhere I can see the code? I am getting ready to make a release because of the broken CQL issue.

**Comment by eplowe** [comment]
I'll post on my fork shortly.

**Comment by eplowe** [comment]
https://github.com/eplowe/fluentcassandra/commits/cqlcoltypes -- I've just included the SetType for now for you to review to make sure you're happy with the direction. I've tested it pretty extensively and seems to be working as expected. Even though in cassandras world a SetType is a java Set and the ListType is a List -- I am thinking about making them the same, but we'll see.

It also needs some additional work IMHO. CompositeType works fine with CassandraObject as a base because it can be a mixed bag of cassandra datatypes. SetType/ListType are just one cassandra datatype, same with the key and value for MapType. Maybe I am over-thinking the problem.

**Comment by nberardi** [comment]
Make sure you get latest. You change includes a change I had already made
to the connection builder.

Looks like a good first start. Please remember that FluentCassandra code
base uses tabs, so that it doesn't throw the formatting off.

One of the things that makes Fluent Cassandra so easy to use is the
transparent types. So make sure that the object can take for example an
array of ints. Something similar was done with the composite types.

I might tackle the additions to CassandraType myself. There are a lot if
points in there that you have to hit. So I just want to make sure they are
all added in the same way.

## 

Nick Berardi
(484) 302-0125
Sent on the go from my phone.

On Jan 5, 2013, at 11:05 PM, Eric Plowe notifications@github.com wrote:

https://github.com/eplowe/fluentcassandra/commits/cqlcoltypes -- I've just
included the SetType for now for you to review to make sure you're happy
with the direction. I've tested it pretty extensively and seems to be
working as expected. Even though in cassandras world a SetType is a java
Set and the ListType is a List -- I am thinking about making them the same,
but we'll see.

—
Reply to this email directly or view it on
GitHubhttps://github.com/managedfusion/fluentcassandra/issues/96#issuecomment-11924123.

**Comment by eplowe** [comment]
Yeah I handled the conversions similar to what you did with composite type.
So you can take an array of objects/CassandraObjects/list/hashset and it'll
handle the conversion back and forth etc.

Yeah I have an add on that is screwing with the formatting I need to
disable it. If you do decide to do it the reading for list is the same as
the set. First two bytes are the number of items. Map is the same except
you do need to read the key out first then the value. I didn't include the
unit test but it seems to behave in line with the Cassandra Type system.

**Comment by eplowe** [comment]
Found out what `EmptyType` is for. Say you create a CF using the API. 

``` c#
keyspace.TryCreateColumnFamily(new CassandraColumnFamilySchema
{
    FamilyName = "Emails",
        KeyValueType = CassandraType.AsciiType,
    ColumnNameType =CassandraType.SetType(CassandraType.AsciiType)
});
```

You'll get from `DESCRIBE KEYSPACE`

```
CREATE TABLE "Emails" (
  "KEY" ascii,
  column1 set<ascii>,
  value blob,
  PRIMARY KEY ("KEY", column1)
) WITH COMPACT STORAGE AND
  bloom_filter_fp_chance=0.010000 AND
  caching='KEYS_ONLY' AND
  comment='' AND
  dclocal_read_repair_chance=0.000000 AND
  gc_grace_seconds=864000 AND
  read_repair_chance=0.100000 AND
  replicate_on_write='true' AND
  compaction={'class': 'SizeTieredCompactionStrategy'} AND
  compression={'sstable_compression': 'SnappyCompressor'};
```

Then using this code:

```
using (var db = new CassandraContext(keyspace: KeyspaceName, server: Server))
{
    var key = "frodo";
    var emailsFamily = db.GetColumnFamily("Emails");

    // create emails
    ConsoleHeader("create emails");

    var emailSet = new SetType(CassandraType.AsciiType);
    emailSet.Add("frodo@gmail.com");
    emailSet.Add("frodo@aol.com");
    emailSet.Add("frodo@live.com");
    emailsFamily.InsertColumn(key, emailSet, Math.PI);
    // attach the emails to the database

    // save the changes
    ConsoleHeader("saving changes");
    db.SaveChanges();
}
```

You'll get the error `EmptyType only accept empty values` back from Cassandra.

**Comment by nberardi** [comment]
Make sure you pull latest changes. I have changed `CassandraType` to support `EmptyType` and `InetAddressType`

**Comment by eplowe** [comment]
Will do. I just need to write a few more unit tests and I'll be pushing up
Set/List/Map.

**Comment by nberardi** [comment]
As I noted with @aaronontheweb please try you best to use tabs instead of spaces when committing.  It looks like your default environment is setup for spaces and not tabs.  Which is no big deal, but there are plugins to help fix the files.  Most notably and the one I use is the Productivity Power Tools for Visual Studio 2012.

http://visualstudiogallery.msdn.microsoft.com/3a96a4dc-ba9c-4589-92c5-640e07332afd

There is also one for Visual Studio 2010. 

http://visualstudiogallery.msdn.microsoft.com/d0d33361-18e2-46c0-8ff2-4adea1e34fef

The name of the tools is 'Fixed Mixed Tabs' which when it detects mixed spacing, it asks you which you would like to use spaces or tabs.  If you could choose Tabs before committing that would be great.

**Comment by eplowe** [comment]
@nberardi will do. I actually switched my environment to tabs and I got the add on over the weekend.

**Comment by nberardi** [comment]
One last place I forgot to mention so that the type conversion works when dealing with the CassandraObject type.

https://github.com/managedfusion/fluentcassandra/blob/master/src/Types/CassandraObject.cs#L176

There are a couple collection types already there, you may need to add ones for Set and Map.  Interfaces are not allowed so these have to be implemented system types.

There is a lot of places to touch to make a type system truly transparent.  :+1:

**Comment by nathannis** [comment]
What is the status of the Collections referenced here?  I am getting FluentCassandra.Operations.CassandraOperationException : Type 'org.apache.cassandra.db.marshal.SetType(org.apache.cassandra.db.marshal.UTF8Type)' not found. when trying to query out a set<text> column.

**Comment by nberardi** [comment]
We are still waiting for a patch to this.  It should be relatively easy to
add, just haven't had the time.

**Comment by Aaronontheweb** [comment]
Forgot to close this issue. It's been resolved since release v1.2.4
